File: 2023/day_8/solution.py
import sys
import math
from functools import reduce

class Instructions:
    def __init__(self, moves):
        self.moves = moves

    def __iter__(self):
        self.p = 0
        return self

    def __next__(self):
        i = self.moves[self.p]
        self.p = (self.p + 1) % (len(self.moves) - 1)
        return i

class Node:
    def __init__(self, name, left, right):
        self.name = name
        self.left = left 
        self.right = right
    
    def move(self, lr):
        if lr == 'R':
            return self.right 
        if lr == 'L':
            return self.left
        else:
            print("bad input")

def getInstructions(line):
    return Instructions(line)

def buildNode(line):
    name, lr = line.split(" = ")
    l, r = lr.replace('\n', '').replace('(', '').replace(')', '').replace(' ', '').split(",")
    return Node(name, l, r)

def distanceToStopNodes(nodes, start, instructions):
    stopDistances = {}
    moves = iter(instructions)
    distance = 0
    node = start
    seen = set()
    while True:
        distance += 1
        m = next(moves)
        node = nodes[node].move(m)
        if (moves.p, node) in seen:
            print("in loop, stopping")
            return stopDistances
        seen.add((moves.p, node))
        if node[-1] == 'Z':
            print("found stop")
            if node not in stopDistances:
                stopDistances[node] = distance
            else:
                print("encountered a duplicate stop. This is likely an error")

def greatestElement(v, values):
    for va in values:
        if v < va:
            return False
    return True

if __name__ == "__main__":
    input = sys.argv[1]
    instructions = None
    with open(input) as f:
        nodes = {}
        for i, l in enumerate(f):
            if i == 0:
                instructions = getInstructions(l)
                continue
            node = buildNode(l)
            nodes[node.name] = node
    
    currentNodes = [x for x in nodes.keys() if 'A' == x[-1]]
    stopNodes = [x for x in nodes.keys() if 'Z' == x[-1]]
    traversals = [distanceToStopNodes(nodes, x, instructions) for x in currentNodes]
    print(currentNodes)
    print(traversals)

    allValues = []
    for t in traversals:
        for _, v in t.items():
            allValues.append(v)
    print(allValues)

    print(math.lcm(traversals))

File: 2023/day_8/solution.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"strings"
)

type Node struct {
	left  string
	right string
}

func getAdjacency(line string) (string, string, string) {
	split := strings.Split(line, " = ")
	adjacencies := strings.Split(split[1], ", ")
	return split[0], adjacencies[0][1:], adjacencies[1][:len(adjacencies[1])-1]
}

func getNextNode(node Node, move string) (string, error) {
	switch move {
	case "L":
		return node.left, nil
	case "R":
		return node.right, nil
	}

	return "", errors.New("unrecognized direction")
}

func getMovesToEnd(adjacencyList map[string]Node, moves string, start string, res chan int) {
	distance := 0
	p := 0
	node := start
	for !strings.Contains(node, "Z") {
		nextNode, error := getNextNode(adjacencyList[node], moves[p:p+1])
		if error != nil {
			fmt.Printf("encountered an error %s", error)
			return
		}

		node = nextNode
		distance += 1
		p = (p + 1) % len(moves)
	}

	res <- distance
}

func GCD(a, b int) int {
	for b != 0 {
		t := b
		b = a % b
		a = t
	}
	return a
}

func LCMForTwo(a, b int) int {
	return a * b / GCD(a, b)
}

func LCM(integers []int) int {
	a := integers[0]
	b := integers[1]
	result := LCMForTwo(a, b)

	for i := 2; i < len(integers); i++ {
		result = LCMForTwo(result, integers[i])
	}

	return result
}

func main() {
	file, _ := os.Open(os.Args[1])
	defer file.Close()

	var adjacencyList map[string]Node = make(map[string]Node)

	var count int = 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		var text string = scanner.Text()
		if count == 0 {
			count += 1
			continue
		}
		name, left, right := getAdjacency(text)
		adjacencyList[name] = Node{left, right}
	}

	moves := os.Args[2]
	var results []chan int
	for node := range adjacencyList {
		if strings.Contains(node, "A") {
			newChan := make(chan int)
			go getMovesToEnd(adjacencyList, moves, node, newChan)
			results = append(results, newChan)
		}
	}

	var nums []int
	for r := range results {
		num := <-results[r]
		nums = append(nums, num)
	}

	fmt.Printf("res of %d", LCM(nums))
}
